docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,650 @@
+of these integers:
+
+- 32 bit attachment generation (this easily accommodates reasonable worst case scenarios like
+  a node restarting every minute for 5 years)
+- 16 bit node ID (this doesn't have to accommodate all pageservers ever, just all the pageservers
+  that act together in a logical cluster)
+- 32 bit node generation (as with attachment generation, this can handle even very frequent restarts)
+
+So an example suffix for generation 1 of node 0 might look like:
+
+```
+   00000001-0000-00000001
+```
+
+This suffix is the primary mechanism for protecting against split-brain situations, and
+enabling safe multi-attachment of tenants:
+
+- Two pageservers running with the same node ID (e.g. after a failure, where there is
+  some rogue pageserver still running) will not try to write to the same objects.
+- Multiple attachments of the same tenant will not try to write to the same objects, as
+  each attachment would have a distinct attachment generation.
+
+### Deletion Part 1: Remote oject deletions
+
+While writes are de-conflicted by writers always using their own generation number in the key,
+deletions are slightly more challenging: if a pageserver A is isolated, and the true active node is
+pageserver B, then it is dangerous for A to do any object deletions, even of objects that it wrote
+itself, because pageserver's B metadata might reference those objects.
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. it is the legitimate holder of its node ID), _and_ that its attachments for the tenants
+  whose data is being deleted are current.
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
+   read metadata that does not reference the about-to-be-deleted object).
+3. Confirm we hold a non-stale generation for the node and the attachment(s) doing deletion.
+4. Actually delete the objects
+
+Because step 3 puts load on the control plane, deletion lists should be executed lazily: write
+them out to S3 and then execute them as a low priority batch operation in the background. Persisting
+deletion lists to S3 is not strictly necessary for correctness (deletions can always be accomplished
+later by object listing & scrub), but provides more flexibility in the case of a given node ID being
+executed on a fresh node with an empty disk: it can recover the deletion queue from the previous
+incarnation of the node ID.
```

Comment:
> object is being uploaded by another PS and about to be added to a metadata file (illegimate delete case)

In this case, a scrubbing pageserver would enqueue the deletion, and then refuse to execute it because the other PS would have been attached with a higher attachment generation number than the one doing the scrubbing (i.e. the validation step in delete would fail).

Added clarification in the scrub section:
>> Scrubbing will be done only by an attached pageserver (not some third party process), and deletions requested during scrub will go through the same
validation as all other deletions: the attachment generation must be
fresh. This avoids the possibility of a stale pageserver incorrectly
thinking than an object written by a newer generation is stale, and deleting
it.

> Separate question on this paragraph: if we persist the deletion queue to S3, is it also suffixed by the generation numbers?

Yes -- not strictly necessary (if two nodes are fighting for the same node ID we could just let them, and accept object leakage), but it is nice for consistency with the overall approach, and doesn't make things much more complicated.

Added clarification in the extended deletion queue description in the optimization section.

fad94685